tools/statplot.py

```diff
@@ -1,229 +1,305 @@
+        self.xData = self.entries[self.xField]
+        self.yData = self.entries[self.yField]
+        if numpy.unique(self.xData).size == 1 \
+                and numpy.unique(self.yData).size == 1:
+            type = 'marker'
+        if action == 'create':
+            self.statplot, = self.ax.plot(self.xData, self.yData, linewidth=2,
+                                          color='xkcd:light purple')
+        elif action == 'update':
+            self.statplot.set_xdata(self.xData)
+            self.statplot.set_ydata(self.yData)
+            if type == 'line':
+                self.statplot.set_color('xkcd:light purple')
+                self.statplot.set_linestyle('solid')
+                self.statplot.set_linewidth(2)
+                self.statplot.set_marker('None')
+            elif type == 'marker':
+                self.statplot.set_linestyle('None')
+                self.statplot.set_marker('d')
+                self.statplot.set_markeredgecolor('xkcd:black')
+                self.statplot.set_markerfacecolor('xkcd:light purple')
+                self.statplot.set_markersize(7)
+                self.statplot.set_markeredgewidth(0.3)
+            self.toolbar.update()
+        self.FormatAxis('x', xscale)
+        self.FormatAxis('y', yscale)
+        self.fig.canvas.draw()
+        self.fig.canvas.flush_events()
+
+    def PopulateCombo(self, action, prevIterX=None, prevIterY=None):
+        for entry in sorted(self.entries.Paths()):
+            self.xCombo.append_text(entry)
+            self.yCombo.append_text(entry)
+        if action == 'load':
+            if 'ElapsedTime' in self.entries.Paths():
+                iterX = self.xCombo.get_model(). \
+                    get_iter(sorted(self.entries.Paths()).index('ElapsedTime'))
+                self.xCombo.set_active_iter(iterX)
+                self.yCombo.set_active(0)
+            else:
+                self.xCombo.set_active(0)
+                self.yCombo.set_active(1)
+        elif action == 'reload':
+            self.xCombo.set_active_iter(prevIterX)
+            self.yCombo.set_active_iter(prevIterY)
+
+    def PopulateCompletion(self, text):
+        return [item for item in sorted(self.entries.Paths())
+                if item.startswith(text)]
+
+    def ReadData(self, statfile):
+        stats = []
+        for i, filename in enumerate(statfile):
+            failcount = 0
+            while failcount < 5:
+                try:
+                    stats.append(fluidity_tools.Stat(filename))
+                    break
+                except (TypeError, ValueError):
+                    time.sleep(0.2)
+                    failcount += 1
+            if failcount == 5:
+                raise Exception('Could not open %s' % filename)
+        if len(stats) == 1:
+            return stats[0]
+        else:
+            return fluidity_tools.JoinStat(*stats)
+
+    def RefreshData(self, statfile):
+        self.xCombo.disconnect_by_func(self.ComboChangedX)
```

Comment:
I think it's more elegant to just block the signal handlers:

```suggestion
        with self.xCombo.handler_block(self.handler_combochangedx), self.yCombo.handler_block(self.handler_combochangedy):
```

This way, the handler unblock is scoped automatically. This requires you to save the handler IDs in the constructor however (`Object.handler_block_by_func` doesn't define `__enter__`, etc.).

Alternatively. you could just block/unblock instead of disconnecting and reconnecting:

```suggestion
        self.xCombo.handler_block_by_func(self.ComboChangedX)
        ...
        self.xCombo.handler_unblock_by_func(self.ComboChangedX)
```

Reply:
Would the following work?

Inside __init__
`self.xCon = self.xCombo.connect('changed', self.ComboChangedX)`
`self.yCon = self.yCombo.connect('changed', self.ComboChangedY)`

`def RefreshData(self, statfile):
        with (self.xCombo.handler_block(self.xCon),
              self.yCombo.handler_block(self.yCon)):
            self.xCombo.remove_all()
            self.yCombo.remove_all()
            self.entries = self.ReadData(statfile)
            self.PopulateCombo('reload',
                               prevIterX=self.xCombo.get_active_iter(),
                               prevIterY=self.yCombo.get_active_iter())`

Reply:
Yes, it should do. Glad you figured out my rather jumbled suggestion!